cmd/get-kustomize-build-base/README.md

```diff
@@ -0,0 +1,24 @@
+# `get-kustomize-build-base`
```

Comment:
Why do we need additional command instead of updating `kustomize-build-dirs`? 
It corresponds to the logic from `kustomize-build-dirs` so much, that it needs function copied from there (findKustomizationRoot). 
It's hard to find the use case  for `get-kustomize-build-base` other than chained with `kustomize-build-dirs` - and it's hard to find the use case for `kustomize-build-dirs` to run without this command.

Reply:
> Why do we need additional command instead of updating `kustomize-build-dirs`? It corresponds to the logic from `kustomize-build-dirs` so much, that it needs function copied from there (findKustomizationRoot). It's hard to find the use case for `get-kustomize-build-base` other than chained with `kustomize-build-dirs` - and it's hard to find the use case for `kustomize-build-dirs` to run without this command.

Reasons for a separate command:

* It's simpler: just adding a bunch of code with clearly defined functionality vs. injecting extra functionality into an existing command (though we could move things around in the future if needed)
* It separates intent: `kustomize-build-dirs` can only worry about running `kustomize build` and this command can worry about finding where to build

> so much, that it needs function copied from there (findKustomizationRoot).

Yeah, ideally I will replace that logic with this logic, but I wanted to see if it works well first

Do you have any concerns with a separate command?

Reply:
> Do you have any concerns with a separate command?

Yes, I believe it greatly increases the cognitive load. 
Do you believe, that making this (already complex command)
 ```
git diff --name-only --raw HEAD^1 HEAD \
            | xargs kustomize-build-dirs --out-dir built-manifests/ --truncate-secrets 2>&1 \
            | tee build-output
```
even more complex, and demanding knowledge of additional component:
```
git diff --name-only --raw HEAD^1 HEAD \
            | get-kustomize-build-base 
            | xargs kustomize-build-dirs --out-dir built-manifests/ --truncate-secrets 2>&1 \
            | tee build-output
```
will simplify the life of the end user?

Or, that creating additional command for sake of running one additional function, which is needed only as an output to another command- will simplify the maintenance of this repository?

Reply:
>It separates intent: kustomize-build-dirs can only worry about running kustomize build and this command can worry about finding where to build

According to this logic, we could propose- for example- splitting the `kustomize-build-dirs` into [following components](https://github.com/utilitywarehouse/manifest-checkers/blob/main/cmd/kustomize-build-dirs/main.go#L62-L99):

- check-kustomize-installed
- find-kustomization-roots
-  remove-component-kustomization

Reply:
Adding it to `kustomize-build-dirs` is non-trivial, since that command expects a bunch of filenames passed as args, which won't work for the output of `git diff --raw`. Since there's a bunch of spaces in each line, I'm not sure we would even know when to start/stop parsing a given line.

My plan is to:

* Slot this command in, see if it really does what we want/gives devs nicer diffs, and if it does
* Come back and tidy up/rethink (e.g. see if we want to throw it all in `kustomize-build-dirs`, and if so make it handle inputs differently etc.)

Reply:
> that command expects a bunch of filenames passed as args, which won't work for the output of git diff --raw. Since there's a bunch of spaces in each line, I'm not sure we would even know when to start/stop parsing a given line.

Aren't you splitting the result of `git diff --raw` in [this PR](https://github.com/utilitywarehouse/manifest-checkers/pull/51/files#diff-5581e059a643c28d8423489ccb67a131431c91fe004b593c19b071de68110a26R70)?

Reply:
>Come back and tidy up/rethink (e.g. see if we want to throw it all in kustomize-build-dirs, and if so make it handle inputs differently etc.)

This is customer- facing tool used in the multiple repositories used by different teams. I think we can't afford half- solutions and experiments for this one. 

If there are cases, where we should keep the high quality of code- it's one of them.

Reply:
> Aren't you splitting the result of git diff --raw in [this PR](https://github.com/utilitywarehouse/manifest-checkers/pull/51/files#diff-5581e059a643c28d8423489ccb67a131431c91fe004b593c19b071de68110a26R70)?

Yeah, I can do that because this command reads from `stdin`, `kustomize-build-dirs` does not and I would like to avoid changing how it takes inputs+adding extra logic in a single change

Reply:
> I would like to avoid changing how it takes inputs+adding extra logic in a single change

Obviously, that's understandable. So, what's your plan?

Reply:
>   Add this new command
    Inject it into the k8s-manifests pipeline as outline above
    Get feedback
    If feedback good, consider merging the logic back into one command or otherwise packaging it nicely so other teams/users can benefit


What are the benefits of this approach?
The only benefit of this approach  which I see, is that we would push the effort into the future. 
It would be simplify the work now, but we would create tech debt*, with hope that "we'll eventually fix it after it will run some time".

*we would have two ways of running the code. And we would have some code replicated in `manifest-checkers`. And we would have mention in the code: "todo: eventually refactor".

Reply:
>    Code quality: should always be high, things is code readability, having readable tests etc.
    Feature quality: how perfect of a product we present to the users, can vary at times (for experiments etc.)

What's the benefit of nice, clean code, if it is doing weird things? 
I think the logic should be the priority before aesthetics